### ISSUE TITLE
Apply to use Flow and kotlinx.Result instead of suspend fun in data layer and custom UseCaseResult

### ISSUE THREAD

**Issue by luongvo** [nimblehq/android-templates#213]
## Issue

Currently, we're using `suspend fun` to transfer the data from `Repository` through `UseCase` to `ViewModel`, then using try/catch to wrap the result which is `Success` or `Fail` in `UseCase` layer. We should migrate this to the new approach by applying [Flow](https://developer.android.com/kotlin/flow).

https://github.com/nimblehq/android-templates/blob/403bd3fa68bfc2c235ff7ccef89973a38b8c341e/CoroutineTemplate/domain/src/main/java/co/nimblehq/coroutine/domain/usecase/GetUsersUseCase.kt#L9-L16

## Solution

- Apply to use [Flow](https://developer.android.com/kotlin/flow) to transfer to transfer to data from `Repository` through `UseCase` to `ViewModel`.
- Here is a quick POC for everyone to checkout how it would be 👉  https://github.com/nimblehq/android-templates/commit/a1b599e9b61a235bc8e4af6cd472179114c915fe. This approach also help us avoid duplicating a new try/catch block in both `UseCase` and `Repository` when implementing https://github.com/nimblehq/android-templates/issues/212 as we can't access `retrofit objects` such as [HttpException](https://square.github.io/retrofit/2.x/retrofit/retrofit2/HttpException.html) in UseCase layer.

## Who Benefits?

Developers

I think this is **essential** because it will help us simplify the error mapping in `Repository` and other components too.

## What's Next?

- [x] Apply to use [Flow](https://developer.android.com/kotlin/flow) instead of `suspend fun`.
- [ ] Apply to use built-in [Result](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/) instead of custom [UseCaseResult](https://github.com/nimblehq/android-templates/blob/develop/CoroutineTemplate/domain/src/main/java/co/nimblehq/coroutine/domain/usecase/UseCaseResult.kt) to wrap `Success` or `Fail` result.
- [x] Update all ViewModel logics to adapt new structure.

**Comment by ryan-conway** [comment]
About using `kotlinx.Result` instead of our own class, do we ever have a use case where we would want to emit another state (for example, `Loading`) before emitting `Success` or `Failure`? 🤔  For example, something like this:
```kotlin
fun execute(): Flow<UseCaseResult<List<Any>>> = flow {
    emit(UseCaseResult.Loading())
    val result = doSomeLongOperation()
    emit(result)
}
```

**Comment by hoangnguyen92dn** [comment]
@ryan-conway I also had the same concern, but in my option, we would need to handle loading state from `ViewModel` instead of `UseCase`, so that's why using the native `Result` class is reasonable 😄

**Comment by ryan-conway** [comment]
@hoangnguyen92dn Fair point 😄  I just figured the in the `ViewModel` it would end up something like 
```kotlin
getUsersUseCase.execute().collect {
    showLoading(result is Loading)
    when (result) {
        is Success -> doSomethingSuccessful(result.data.toUserUiModels())
        is Failure -> handleFailure(result.error.message.orEmpty())
    }
}
```
instead of 
```kotlin
getUsersUseCase.execute().collect {
    showLoading()
    if (result.isSuccess) doSomethingSuccessful(result.getOrNull()!!.toUserUiModels())
    else handleFailure(result.exceptionOrNull()!!.message.orEmpty())
    hideLoading()
}
```
But if the idea is to have `ViewModel` fully handle loading then I suppose `kotlinx.Result` makes sense to use 😄

**Comment by luongvo** [comment]
@ryan-conway From your example

```
getUsersUseCase.execute().collect {
    showLoading(result is Loading)
    when (result) {
        is Success -> doSomethingSuccessful(result.data.toUserUiModels())
        is Failure -> handleFailure(result.error.message.orEmpty())
    }
}
```

This approach is so hard to control the loading state when having **multiple tasks in parallel**. The early finishing task will dismiss the loading while others are still running. That's why we designed the control the loading in `BaseViewModel` by counting the tasks for loading showing and hiding to simplify the loading controlling logic in ViewModel.

https://github.com/nimblehq/android-templates/blob/403bd3fa68bfc2c235ff7ccef89973a38b8c341e/CoroutineTemplate/app/src/main/java/co/nimblehq/coroutine/ui/base/BaseViewModel.kt#L28-L46

By that way, we don't need to control loading state in the Result model. Also from my experiences with our projects in the past, we don't have extra states in the Result model too.

**Comment by ryan-conway** [comment]
@luongvo Also a fair point, so yeah `kotlinx.Result` sounds good then 😄

**Comment by luongvo** [comment]
@Tuubz I would like to delegate this task back to @manh-t 

FYI, here is more insights of Flow applying in our running project. @manh-t please check it out some more details:
- The need of applying Flow to replace suspend fun to build call chain with `map`, `flatMap` or `error catching` flow.

  ![image](https://user-images.githubusercontent.com/16315358/169945361-e85216eb-5a34-41bc-9d85-5b3ef8dadb89.png)

- The need of applying `kotlin.Result`. **I think we don't need a result model anymore.**
- Instead of throwing the error directly in Repository suspend fun, we can throw the error inside of Flow stream. Custom errors catching from network layer could be executed in our custom error mapping extension instead of using `try/catch` in each UseCase.

  <img width="1258" alt="image" src="https://user-images.githubusercontent.com/16315358/169945606-27631b57-7b92-4d6e-9a47-dd07252f7604.png">

  <img width="597" alt="image" src="https://user-images.githubusercontent.com/16315358/169945667-faed3073-5ecb-4c6c-bb62-9346c18f31a9.png">

Feedback is welcome before @manh-t starts working his contribution 👋

**Comment by Tuubz** [comment]
Voting complete 🚀

**Comment by luongvo** [comment]
@Tuubz I think we can remove `RFC:` prefix from the title 😄

**Comment by ryan-conway** [comment]
Issue was closed too early 🙈 Re-opening until #292 is merged 👋🏻
